pkg/storage/watch_cache.go

```diff
@@ -250,6 +250,20 @@ func (w *watchCache) GetAllEventsSinceThreadUnsafe(resourceVersion uint64) ([]wa
 	if size > 0 {
 		oldest = w.cache[w.startIndex%w.capacity].resourceVersion
 	}
+	if resourceVersion == 0 {
+		// resourceVersion = 0 means that we don't require any specific starting point
+		// and we would like to start watching from ~now.
+		// However, to keep backward compatibility, we additionally need to return the
+		// current state and only then start watching from that point.
+		//
+		// TODO: In v2 api, we should stop returning the current state - #13969.
+		allItems := w.store.List()
```

Comment:
It is safe (due to how it is implemented).
Basically, the backing store contains a correct state at some point in time. The way it is implemented is that the cacher is locking watchCache, gets the state from that point in time (using list on the backing store) and starts watch from exactly that point. Every new watch event delivered to the backing store will be propagated to that watcher,
So there is no risk that some event will not be delivered.